elasticsearch/sandboxes/heka/output/elasticsearch_bulk_api.lua

```diff
@@ -107,9 +107,24 @@ local function send_request() -- hand coded since socket.http doesn't support ke
             local ok, doc = pcall(rjson.parse, response)
             if ok then
                 if doc:value(doc:find("errors")) then
-                    print(string.format("ElasticSearch server reported errors processing the submission, not all messages were indexed"))
-                    -- todo track partial batch failure counts https://github.com/mozilla-services/lua_sandbox_extensions/issues/89
-                    -- the partial failure is most likely due to bad input, so no retry is attempted as it would just fail again
+                    local resp_error_count = 0
+                    local items = doc:find("items")
+                    for _,i in doc:iter(items) do
+                        for _,a in doc:iter(i) do
+                            if (doc:value(doc:find(a, "status")) or 0) >= 400 then
+                                resp_error_count = resp_error_count + 1
+                            end
+                        end
+                    end
+
+                    if resp_error_count >= batch_count then
```

Comment:
The assumption that if all items failed it must be a cluster problem will most like be incorrect under some conditions so we will inevitably error out when we should have ignored.  There isn't really anything useful in the elastic search API that allows us to easily determine if we should retry or not since there isn't documentation of the various types of errors (we would have to troll the source and manually maintain the set of fatal vs retry errors); which is why the default of log an error and skip was chosen.  I would be a -1 on changing the default behavior without a clear win.

Reply:
Would you accept this change if I restrict it to 403 status ?

Reply:
That would at least address the bad data error vs retry.  What other status codes does it return on a per item bases and under what conditions?

Reply:
no idea, and haven't found docs yet :(

Reply:
https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/rest/RestStatus.java
https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java#L139
I would say 401 and 403 could be retriable without side effect (it will not create duplicates or ...)

Reply:
I would suspect you would not see a 401 at the item level

Reply:
yep, just pushed an updated commit